backend/msgpack/placement.go

```diff
@@ -0,0 +1,269 @@
+	"github.com/m3db/m3x/clock"
+)
+
+var (
+	errNilPlacementSchema            = errors.New("nil placement schema")
+	errNilPlacementSnapshotsSchema   = errors.New("nil placement snapshots schema")
+	errNoApplicablePlacementSnapshot = errors.New("no applicable placement snapshot found")
+	errPlacementClosed               = errors.New("placement is closed")
+)
+
+// placement stores placement information.
+type placement interface {
+	// NumShards returns the number of shards.
+	NumShards() (int, error)
+
+	// Shard returns the shard associated with a given metric id.
+	Shard(id []byte) (uint32, error)
+
+	// RouteShard routes msgpack encoded buffer for a given shard.
+	RouteShard(shard uint32, buf msgpack.Buffer) error
+
+	// Close closes the topology.
+	Close() error
+}
+
+type activePlacement struct {
+	sync.RWMutex
+
+	snapshots []*placementSnapshot
+	nowFn     clock.NowFn
+	queueMgr  instanceQueueManger
+	expiring  int32
+	closed    bool
+}
+
+func newActivePlacement(snapshots []*placementSnapshot, opts placementOptions) placement {
+	queueMgr := opts.InstanceQueueManger()
+	for _, snapshot := range snapshots {
+		queueMgr.AddInstances(snapshot.Instances())
+	}
+	return &activePlacement{
+		snapshots: snapshots,
+		nowFn:     opts.ClockOptions().NowFn(),
+		queueMgr:  queueMgr,
+	}
+}
+
+func (p *activePlacement) NumShards() (int, error) {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return 0, err
+	}
+	numShards := snapshot.NumShards()
+	p.RUnlock()
+	return numShards, nil
+}
+
+func (p *activePlacement) Shard(id []byte) (uint32, error) {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return 0, err
+	}
+	shard := snapshot.Shard(id)
+	p.RUnlock()
+	return shard, nil
+}
+
+func (p *activePlacement) RouteShard(shard uint32, buf msgpack.Buffer) error {
+	p.RLock()
+	snapshot, err := p.activeSnapshotWithLock(p.nowFn())
+	if err != nil {
+		p.RUnlock()
+		return err
+	}
+	instances := snapshot.InstancesForShard(shard)
+	err = p.queueMgr.WriteTo(instances, buf)
+	p.RUnlock()
+	return err
+}
+
+func (p *activePlacement) Close() error {
+	p.Lock()
+	defer p.Unlock()
+
+	if p.closed {
+		return errPlacementClosed
+	}
+	for _, snapshot := range p.snapshots {
+		p.queueMgr.RemoveInstances(snapshot.Instances())
+	}
+	p.snapshots = nil
+	return nil
+}
+
+func (p *activePlacement) activeSnapshotWithLock(t time.Time) (*placementSnapshot, error) {
+	if p.closed {
+		return nil, errPlacementClosed
+	}
+	idx := p.activeIndexWithLock(t)
+	if idx < 0 {
+		return nil, errNoApplicablePlacementSnapshot
+	}
+	snapshot := p.snapshots[idx]
+	// If the snapshot that's in effect is not the first placment, expire the stale ones.
+	if idx > 0 && atomic.CompareAndSwapInt32(&p.expiring, 0, 1) {
+		go p.expire()
+	}
+	return snapshot, nil
+}
+
+func (p *activePlacement) activeIndexWithLock(t time.Time) int {
+	timeNs := t.UnixNano()
+	idx := len(p.snapshots) - 1
```

Comment:
is it more likely for the active snapshot to be in the front of the list? it should be either the 0th or 1st item?